File: leetcode/1673-find-the-most-competitive-subsequence/mostCompetitive.ts
function mostCompetitive(nums: number[], k: number): number[] {
  const result: number[] = [];
  for (const [index, num] of nums.entries()) {
    while (result.length > 0 && result[result.length - 1] > num) {
      if (nums.length - index + 1 + result.length - 1 > k) {
        result.pop();
      } else {
        break;
      }
    }

    if (result.length < k) {
      result.push(num);
    }
  }

  return result;
}

export default mostCompetitive;

File: leetcode/1673-find-the-most-competitive-subsequence/mostCompetitive.rs
use crate::Solution;

impl Solution {
    pub fn most_competitive(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut result: Vec<i32> = Vec::new();
        for (index, num) in nums.iter().enumerate() {
            while result.len() > 0 && result.last().unwrap() > num {
                if (nums.len() - index + 1 + result.len() - 1) as i32 > k {
                    result.pop();
                } else {
                    break;
                }
            }
            if (result.len() as i32) < k {
                result.push(*num)
            }
        }

        return result;
    }
}
